YOU SHOULD PROBABLY RUN THIS ON GOOGLE COLAB!!!

In [1]:
!pip install torch transformers datasets pandas peft

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 90.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 68.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 49.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 10.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 14.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 10.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 94.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 485.4/485.4 kB 29.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 10.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

In [2]:
import pandas as pd
import torch
import ast
from peft import LoraConfig, get_peft_model, PeftModel
import transformers
from transformers import AutoModelForCausalLM, AutoTokenizer, Trainer, TrainingArguments
from datasets import load_dataset, Dataset
import random

In [3]:


df = pd.read_csv("chess_challenges_train_10k.csv")
df["Move"] = df["Move"].apply(ast.literal_eval)  # Ensures it's a real list
df["Win Probability"] = df["Win Probability"].apply(ast.literal_eval)  # Convert string to list of floats

# Now convert Win Probability list to a list of strings
df["Win Probability"] = df["Win Probability"].apply(lambda x: [str(prob) for prob in x])
dataset = Dataset.from_pandas(df)

df = pd.read_csv("chess_challenges_test_2k.csv")
df["Move"] = df["Move"].apply(ast.literal_eval)  # Ensures it's a real list
df["Win Probability"] = df["Win Probability"].apply(ast.literal_eval)  # Convert string to list of floats

# Now convert Win Probability list to a list of strings
df["Win Probability"] = df["Win Probability"].apply(lambda x: [str(prob) for prob in x])
dataset_eval = Dataset.from_pandas(df)


In [4]:
print(dataset_eval[4])

{'FEN': '8/5k2/1pp3pp/3p4/1P2bP1P/P1R3P1/5K2/8 b - - 1 32', 'Move': ['d5d4', 'b6b5', 'c6c5', 'f7e7', 'g6g5', 'e4b1', 'e4d3', 'f7g8', 'h6h5', 'f7e8', 'f7f8', 'f7g7', 'e4f5', 'e4h1', 'f7f6', 'f7e6', 'e4g2', 'e4f3', 'e4c2'], 'Win Probability': ['0.14358734593333494', '0.11965646568318677', '0.11507973387664883', '0.10814488465274963', '0.10779026169080513', '0.10638197729176951', '0.1056839310673574', '0.10533642481155803', '0.10464443579540594', '0.10361398030487234', '0.10157996710955025', '0.10057629832289083', '0.09958144717646433', '0.09697119469294446', '0.09285968753405982', '0.09224121164974286', '0.07688776954513865', '0.0718211393064348', '0.06822522266416714']}


In [5]:
# DATASET construction
from collections import defaultdict

def fen_to_description(fen: str) -> str:
    """
    Converts a FEN (Forsyth-Edwards Notation) string into a human-readable chessboard description.

    Args:
        fen (str): The FEN string representing the board state.

    Returns:
        str: A formatted description of the board state.
    """
    piece_map = {
        "K": "King",
        "Q": "Queen",
        "R": "Rook",
        "B": "Bishop",
        "N": "Knight",
        "P": "Pawn",
        "k": "King",
        "q": "Queen",
        "r": "Rook",
        "b": "Bishop",
        "n": "Knight",
        "p": "Pawn",
    }

    try:
        fen_parts = fen.split()
        if len(fen_parts) < 2:
            raise ValueError(
                "Invalid FEN format. Ensure it has at least a board position and turn information."
            )

        ranks = fen_parts[0].split("/")
        if len(ranks) != 8:
            raise ValueError("Invalid FEN format. The board should have 8 ranks.")

        turn = "White to move." if fen_parts[1] == "w" else "Black to move."
        board = []

        for r, rank in enumerate(ranks):
            row = []
            file = 0
            for char in rank:
                if char.isdigit():
                    file += int(char)
                elif char in piece_map:
                    row.append((char, file, 8 - r))  # (Piece, File, Rank)
                    file += 1
                else:
                    raise ValueError(f"Invalid character '{char}' in FEN notation.")
            board.extend(row)

        piece_positions = defaultdict(list)

        for piece, file, rank in board:
            color = "White" if piece.isupper() else "Black"
            piece_type = piece_map[piece]
            position = f"{chr(file + 97)}{rank}"
            piece_positions[(color, piece_type)].append(position)

        description = [turn]

        for (color, piece_type), positions in sorted(
            piece_positions.items(), key=lambda x: (x[0][0], x[0][1])
        ):
            position_text = ", ".join(positions)
            description.append(
                f"{color} {piece_type}{'s' if len(positions) > 1 else ''} on {position_text}."
            )

        return "\n".join(description)

    except ValueError as e:
        return f"Error processing FEN: {e}"
    except Exception as e:
        return f"Unexpected error: {str(e)}"

def extract_best_move(example):
    # Tokenize the question
    moves = example["Move"][0]
    random.shuffle(example["Move"])
    board = fen_to_description(example["FEN"]) + "\nValid moves: " + str(example["Move"])

    # Tokenize the label (treat as text)
    # label_tokens = tokenizer(example["label"], padding=False, truncation=True)

    return {
        "Board": board,
        "Move": moves,
    }

processed_dataset = dataset.map(extract_best_move)
processed_dataset_eval = dataset_eval.map(extract_best_move)

Map:   0%|          | 0/5000 [00:00<?, ? examples/s]

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

In [7]:
print(processed_dataset[100]["Board"])

White to move.
Black Bishops on b8, f5.
Black King on e8.
Black Knights on c6, f6.
Black Pawns on b7, f7, g7, h7, a6, e6, d5.
Black Queen on d8.
Black Rooks on a8, h8.
White Bishops on b2, e2.
White King on g1.
White Knights on b5, f3.
White Pawns on c4, d4, b3, a2, f2, g2, h2.
White Queen on d1.
White Rooks on a1, f1.
Valid moves: ['f3h4', 'd1d3', 'b3b4', 'b2c1', 'b5d6', 'g1h1', 'd1c1', 'f1e1', 'd1b1', 'a1b1', 'd1c2', 'f3e5', 'f3d2', 'a2a3', 'b5a7', 'e2d3', 'f3g5', 'h2h3', 'h2h4', 'c4d5', 'd1d2', 'g2g4', 'b5c3', 'a2a4', 'b2a3', 'f3e1', 'b2c3', 'c4c5', 'a1c1', 'g2g3', 'b5a3', 'd1e1', 'b5c7']


In [5]:
# from transformers import DataCollatorForLanguageModeling

# data_collator = DataCollatorForLanguageModeling(
#     tokenizer=tokenizer,
#     mlm=False  # No masked language modeling for causal LMs
# )

# train_dataloader = DataLoader(tokenized_dataset, batch_size=1, shuffle=True, collate_fn=data_collator)


In [8]:
def preprocess_function(example):
    prompt = f"<answer> {example['Move']} </answer>"
    input_text = example["Board"]

    tokenized = tokenizer(
        input_text + "\n" + prompt,
        truncation=True,
        padding="max_length",
        max_length=512
    )

    labels = tokenized["input_ids"].copy()
    answer_tokens = tokenizer.encode("<answer>", add_special_tokens=False)

    if answer_tokens[0] in tokenized["input_ids"]:
        answer_start = tokenized["input_ids"].index(answer_tokens[0])
        labels[:answer_start] = [-100] * answer_start

    tokenized["labels"] = labels
    assert len(tokenized["labels"]) == len(tokenized["input_ids"]), "Mismatch in tokenized label length!"

    return tokenized

In [9]:
# ---- 1. Load Model & Tokenizer ----
tokenizer = AutoTokenizer.from_pretrained("deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B")
model = AutoModelForCausalLM.from_pretrained("deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B")

lora_config = LoraConfig(
        r=128,
        lora_alpha=32,
        lora_dropout=0.05,
        bias="none",
        task_type="CAUSAL_LM"
    )

model = get_peft_model(model, lora_config)

# ---- 2. Prepare Dataset ----
tokenized_dataset = processed_dataset.map(preprocess_function, batched=False)
tokenized_dataset_eval = processed_dataset_eval.map(preprocess_function, batched=False)

# ---- 3. Training Arguments ----
training_args = TrainingArguments(
    output_dir="./chess_llm_finetuned",
    per_device_train_batch_size=4,
    per_device_eval_batch_size=4,
    gradient_accumulation_steps=8,
    num_train_epochs=1,
    optim="adamw_torch",
    learning_rate=2e-5,
    weight_decay=0.01,
    save_strategy="epoch",
    eval_strategy="epoch",
    logging_steps=10,
    bf16=True,  # Use bfloat16 for memory efficiency
    push_to_hub=False,  # Set to True if using Hugging Face Hub
    report_to="wandb",  # Log to Weights & Biases
    fp16=False,  # bfloat16 preferred over fp16 for stability
)

# ---- 4. Trainer API ----
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset,
    eval_dataset=tokenized_dataset_eval,
    tokenizer=tokenizer,
)

# ---- 5. Train the Model ----
trainer.train()

# ---- 6. Save Model ----
trainer.save_model("./chess_llm_finetuned")
tokenizer.save_pretrained("./chess_llm_finetuned")


/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/3.07k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/7.03M [00:00<?, ?B/s]

config.json:   0%|          | 0.00/679 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/3.55G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/181 [00:00<?, ?B/s]

Map:   0%|          | 0/5000 [00:00<?, ? examples/s]

Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

<ipython-input-9-786f8ea87a45>:39: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(
wandb: WARNING The `run_name` is currently set to the same value as `TrainingArguments.output_dir`. If this was not intended, please specify a different run name by setting the `TrainingArguments.run_name` parameter.
wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter:

 ··········


wandb: WARNING If you're specifying your api key in code, ensure this code is not shared publicly.
wandb: WARNING Consider setting the WANDB_API_KEY environment variable, or running `wandb login` from the command line.
wandb: No netrc file found, creating one.
wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: lmitbo (lukemitbo) to https://api.wandb.ai. Use `wandb login --relogin` to force relogin


('./chess_llm_finetuned/tokenizer_config.json',
 './chess_llm_finetuned/special_tokens_map.json',
 './chess_llm_finetuned/tokenizer.json')

The purpose of the next few cells is to save the model locally (allowing you to disconnect from A100 and reconnect to a T4 so that you don't burn credits) and then you can upload the model once you reconnect

In [7]:
!zip -r chess_llm_finetuned.zip ./chess_llm_finetuned

  adding: chess_llm_finetuned/ (stored 0%)
  adding: chess_llm_finetuned/README.md (deflated 66%)
  adding: chess_llm_finetuned/training_args.bin (deflated 51%)
  adding: chess_llm_finetuned/tokenizer.json (deflated 81%)
  adding: chess_llm_finetuned/tokenizer_config.json (deflated 85%)
  adding: chess_llm_finetuned/adapter_model.safetensors (deflated 9%)
  adding: chess_llm_finetuned/checkpoint-156/ (stored 0%)
  adding: chess_llm_finetuned/checkpoint-156/trainer_state.json (deflated 72%)
  adding: chess_llm_finetuned/checkpoint-156/README.md (deflated 66%)
  adding: chess_llm_finetuned/checkpoint-156/training_args.bin (deflated 51%)
  adding: chess_llm_finetuned/checkpoint-156/tokenizer.json (deflated 81%)
  adding: chess_llm_finetuned/checkpoint-156/tokenizer_config.json (deflated 85%)
  adding: chess_llm_finetuned/checkpoint-156/optimizer.pt (deflated 7%)
  adding: chess_llm_finetuned/checkpoint-156/adapter_model.safetensors (deflated 9%)
  adding: chess_llm_finetuned/checkpoint-15

In [8]:
from google.colab import files
files.download("chess_llm_finetuned.zip")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [1]:
!unzip chess_llm_finetuned.zip -d ./chess_llm_finetuned

Archive:  chess_llm_finetuned.zip
   creating: ./chess_llm_finetuned/chess_llm_finetuned/
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/README.md  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/training_args.bin  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/tokenizer.json  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/tokenizer_config.json  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/adapter_model.safetensors  
   creating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/trainer_state.json  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/README.md  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/training_args.bin  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/tokenizer.json  
  inflating: ./chess_llm_finetuned/chess_llm_finetuned/checkpoint-156/tokenizer_config.json  
  inflating: ./ches

In [10]:
# Load base model first
model = AutoModelForCausalLM.from_pretrained("deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B")
# Load fine-tuned LoRA weights
model = PeftModel.from_pretrained(model, "./chess_llm_finetuned")
model = model.merge_and_unload()  # Only if you want to merge weights permanently
tokenizer = AutoTokenizer.from_pretrained("./chess_llm_finetuned")

# ---- CONSTRAINED DECODING FOR DEEPSEEK TOKENIZER ----
def generate_move_samples(
    english_description: str,
    valid_moves: list[str],
    model,
    tokenizer,
    k: int = 5
) -> list[str]:
    """
    Fix for subword tokenization. Constrains decoding to only tokens
    matching parts of the valid UCI moves.
    """
    # Tokenize the initial text (include <answer> tag)
    prompt = f"{english_description}\n<answer>"
    input_ids = tokenizer.encode(prompt, return_tensors="pt")

    samples = []
    for _ in range(k):
        current_ids = input_ids.clone()
        generated_tokens = []

        # Generate until </answer> or max 7 tokens (UCI-move length + safety)
        while True:
            # Get logits for next token
            with torch.no_grad():
                outputs = model(current_ids)
            logits = outputs.logits[:, -1, :]

            # Get allowed tokens: Any token that might continue a valid move
            allowed_token_ids = []
            partial_sequence = tokenizer.decode(generated_tokens)
            for move in valid_moves:
                if move.startswith(partial_sequence):
                    possible_continuation = tokenizer.encode(
                        move[len(partial_sequence):],
                        add_special_tokens=False
                    )
                    if possible_continuation:
                        allowed_token_ids.append(possible_continuation[0])

            # Handle case with no allowed tokens (broken FEN/valid_moves mismatch)
            if not allowed_token_ids:
                break

            # Mask out invalid tokens and sample
            mask = torch.ones_like(logits) * -float("inf")
            mask[:, allowed_token_ids] = 0
            masked_logits = logits + mask

            probs = torch.softmax(masked_logits, dim=-1)
            next_token_id = torch.multinomial(probs, num_samples=1).item()
            generated_tokens.append(next_token_id)
            current_ids = torch.cat([current_ids, torch.tensor([[next_token_id]])], dim=-1)

            # Check if generated sequence is a valid move
            current_move = tokenizer.decode(generated_tokens, skip_special_tokens=True)
            if current_move in valid_moves:
                samples.append(current_move)
                break

            # Safety break to prevent infinite loops
            if len(generated_tokens) >= 7:
                break

    return samples

# Example usage
def evaluate_position(fen: str, valid_moves: list[str], k: int = 5):
    description = fen_to_description(fen)
    samples = generate_move_samples(description, valid_moves, model, tokenizer, k=k)

    from collections import defaultdict
    counts = defaultdict(int)
    for move in samples:
        counts[move] += 1
    mode = max(counts, key=counts.get, default=None)

    return mode

# Test with your dataset entry
example = processed_dataset_eval[0]
print(example)
valid_moves = ast.literal_eval(example["Board"][example["Board"].find("[") : example["Board"].find("]")+1])  # Ensure this is the full list, not just the best move!
print(valid_moves)

predicted_move = evaluate_position(
    fen=example["FEN"],
    valid_moves=valid_moves,
    k=5
)
print(f"True move: {example['Move']}")
print(f"Predicted move: {predicted_move}")


{'FEN': 'rnbq1rk1/1p2ppbp/p4np1/2p5/8/2NBPN2/PPPB1PPP/R2QR1K1 b - - 1 9', 'Move': 'b8c6', 'Win Probability': ['0.5477213400816676', '0.53675438329054', '0.5257517538936253', '0.5257517538936253', '0.5156437324214209', '0.5073636275599663', '0.5036820134406647', '0.5018410316799476', '0.5', '0.49907948104001076', '0.49907948104001076', '0.49907948104001076', '0.4981589683200522', '0.49539752999711567', '0.4926363724400336', '0.47516641034760715', '0.46874313267512907', '0.467826316589383', '0.4669097177008696', '0.4650771960458313', '0.46416128552924907', '0.46416128552924907', '0.4605001214877426', '0.44226590014833245', '0.44226590014833245', '0.43501278527292314', '0.43501278527292314', '0.4277873846879599', '0.4259857125367345', '0.4027589492090241', '0.33027105825721115', '0.1708503939715894', '0.14044687122292038', '0.09991208792971568', '0.07610740636397961', '0.040442586480336384'], 'Board': "Black to move.\nBlack Bishops on c8, g7.\nBlack King on g8.\nBlack Knights on b8, f6.\n

I have no idea if the above is working rn but at least it's not giving any errors.

b8c6: 0.5477213400816676
g6g5 0.33027105825721115

not good.